### DISCUSSION TITLE
type-conversions/others

### DISCUSSION THREAD

**Discussion by giscus** [sunface/rust-by-practice#176] [Book Comments]
# type-conversions/others

Learning Rust By Practice, narrowing the gap between beginner and skilled-dev with challenging examples, exercises and projects.

https://zh.practice.rs/type-conversions/others.html

**Comment by jingth** [comment]
/* 
use std::fmt::{self, Display};

struct Point {
    x: i32,
    y: i32,
}

impl fmt::Display for Point {
    // 实现 fmt 方法
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = format!("The point is ({}, {})", self.x, self.y);
        f.write_str(&s)
    }
}

fn main() {
    let origin = Point { x: 0, y: 0 };
    // 填空
    assert_eq!(origin.to_string(), "The point is (0, 0)");
    assert_eq!(format!("{}", origin), "The point is (0, 0)");

    println!("Success!")
}
 */

/* 
 // 为了使用 `from_str` 方法, 你需要引入该特征到当前作用域中
use std::str::FromStr;

fn main() {
    let parsed: i32 = i32::from_str("5").unwrap();
    let turbo_parsed = i32::from_str("10").unwrap();
    let from_str: i32 = i32::from_str("20").unwrap();
    let sum = parsed + turbo_parsed + from_str;
    assert_eq!(sum, 35);

    println!("Success!")
}
 */


use std::str::FromStr;
use std::num::ParseIntError;

#[derive(Debug, PartialEq)]
struct Point {
    x: i32,
    y: i32
}

impl FromStr for Point {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let coords: Vec<&str> = s.trim_matches(|p| p == '(' || p == ')' )
                                 .split(',')
                                 .collect();

        let x_fromstr = coords[0].parse::<i32>()?;
        let y_fromstr = coords[1].parse::<i32>()?;

        Ok(Point { x: x_fromstr, y: y_fromstr })
    }
}
fn main() {
    // 使用两种方式填空
    // 不要修改其它地方的代码
    let p = Point::from_str("(3,4)"); //3,4之间不能有空格
    assert_eq!(p.unwrap(), Point{ x: 3, y: 4} );

    println!("Success!")
}

**Comment by ADkun** [comment]
### 第1题
```rust
use std::fmt;

struct Point {
    x: i32,
    y: i32,
}

impl fmt::Display for Point {
    // 实现 fmt 方法
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "The point is ({}, {})", self.x, self.y)
    }
}

fn main() {
    let origin = Point { x: 0, y: 0 };
    // 填空
    assert_eq!(origin.to_string(), "The point is (0, 0)");
    assert_eq!(format!("{}", origin), "The point is (0, 0)");

    println!("Success!")
}
```

### 第2题
```rust
// 为了使用 `from_str` 方法, 你需要引入该特征到当前作用域中
use std::str::FromStr;
fn main() {
    let parsed: i32 = "5".parse().unwrap();
    let turbo_parsed: i32 = "10".parse().unwrap();
    let from_str = i32::from_str("20").unwrap();
    let sum = parsed + turbo_parsed + from_str;
    assert_eq!(sum, 35);

    println!("Success!")
}
```

### 第3题
```rust
use std::str::FromStr;
use std::num::ParseIntError;

#[derive(Debug, PartialEq)]
struct Point {
    x: i32,
    y: i32
}

impl FromStr for Point {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let coords: Vec<&str> = s.trim_matches(|p| p == '(' || p == ')' )
                                 .split(',')
                                 .collect();

        let x_fromstr = coords[0].parse::<i32>()?;
        let y_fromstr = coords[1].parse::<i32>()?;

        Ok(Point { x: x_fromstr, y: y_fromstr })
    }
}
fn main() {
    // 使用两种方式填空
    // 不要修改其它地方的代码
    let p = Point::from_str("(3,4)");
    assert_eq!(p.unwrap(), Point{ x: 3, y: 4} );

    println!("Success!")
}
```

**Reply by fengzhongmanbu** [reply]
3题第二种方式：
let p = "(3,4)".parse::<Point>();

**Reply by fengzhongmanbu** [reply]
错了，是 let p = "(3,4)".parse::\<Point\>();

**Comment by Bruce012356** [comment]
mark

**Reply by runzhi214** [reply]
# 第一题
实现Display自动实现to_string的确是学习了。
```rust
use std::fmt;
use std::fmt::Formatter;

struct Point {
    x: i32,
    y: i32,
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f,"The point is ({}, {})",self.x,self.y)
    }
}

fn main() {
    let origin = Point { x: 0, y: 0 };
    // 填空
    assert_eq!(origin.to_string(), "The point is (0, 0)");
    assert_eq!(format!("{}",origin), "The point is (0, 0)");

    println!("Success!")
}
```

**Comment by chenhai998** [comment]
done

**Comment by linkedlist771** [comment]
done

**Comment by TDAkory** [comment]
mark finished

**Comment by bansal-harsh-2504** [comment]
Day 10
done✅
